File: algorithms/0121.best-time-to-buy-and-sell-stock/solution.rs
impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let mut max_profit = 0i32;
        let mut buy_price = i32::MAX;

        for price in prices {
            buy_price = buy_price.min(price);
            max_profit = max_profit.max(price - buy_price);
        }

        max_profit
    }
}

File: algorithms/0121.best-time-to-buy-and-sell-stock/solution.ts
function maxProfit(prices: number[]): number {
  let max = 0;
  let buyPrice = prices[0];

  for (const price of prices) {
    if (price < buyPrice) {
      buyPrice = price;
    }

    if (max < price - buyPrice) {
      max = price - buyPrice;
    }
  }

  return max;
}
